tests/cases/compiler/comparisonCountLimits.ts

```diff
@@ -0,0 +1,27 @@
+// @noTypesAndSymbols: true
+
+function get11BitsOf<T extends string, U extends string, FallbackPrefix extends string>() {
+    type Bit = T | U; // 2^1
+    type HalfNibble = `${Bit}${Bit}`; // 2^2
+    type Nibble = `${HalfNibble}${HalfNibble}`; // 2^4
+    type Byte = `${Nibble}${Nibble}`; // 2^8
+    type ElevenBits = `${Byte}${HalfNibble}${Bit}`; // 2^11 (approx. 4 million comparisons if compared naively)
+
+    type HalfWord = `${Byte}${Byte}`; // 2^16 // allowed, but test takes way too long if used
+    type Word = `${HalfWord}${HalfWord}`; // 2^32 (throws, too large)
+
+    // Literal type relations are uncached (lol), so everything has to be wrapped in an object to affect cache sizes
+    // (A distributive conditional is the easiest way to do the mapping, but others are possible, eg, mapped types,
+    // or explicit construction)
+    type Box<T> = T extends unknown ? {item: T} : never;
+
+    // By manufacturing the fallback in here, we guarantee it has a higher typeid than the bit strings,
+    // and thus is sorted to the end of the union, guaranteeing relationship checking passes with a maximal
+    // number of comparisons when a naive comparison is done (guaranteeing this test is slow)
+    return null as any as Box<ElevenBits | (FallbackPrefix extends never ? never : `${FallbackPrefix}${string}`)>; // return type is a union

```

Comment:
As a double-aside, while writing this speedrun-to-slow-the-compiler-down test, I've also come up with some heuristics for union comparisons that could probably make this test specifically complete in linear time/comparisons.

Specifically, the "union of specific things with a generalized fallback member" pattern is very common. If we store the first target member that passes and check it first for subsequent source members, it can shortcut iteration in these scenarios and prevent a quadratic search. Similar element-index-carryover could also be used to handle corresponding-unions-but-missing-a-element better, like comparing `A | B | D` with `A | B | C | D`, or, more generally, any filtered union with the original. I'll probably put up a PR for that at some point.

Reply:
Is that the same / similar to #47511?

Reply:
I don't think so, since that's talking about conditional type distribution and I'm just talking about relating two unions. Specifically, in that case, it's about how once you have selected both branch types (assuming they don't use the check type in them), you can stop distributing, since further results won't change the outcome. Similar naive complexity, though.